sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/_transport.py

```diff
@@ -523,44 +523,51 @@ def _wrap_socket_sni(  # pylint: disable=no-self-use
         # Setup the right SSL version; default to optimal versions across
         # ssl implementations
         if ssl_version is None:
-            ssl_version = ssl.PROTOCOL_TLS
+            ssl_version = ssl.PROTOCOL_TLS_SERVER if server_side else ssl.PROTOCOL_TLS_CLIENT
 
         opts = {
             "sock": sock,
-            "keyfile": keyfile,
-            "certfile": certfile,
             "server_side": server_side,
-            "cert_reqs": cert_reqs,
-            "ca_certs": ca_certs,
             "do_handshake_on_connect": do_handshake_on_connect,
             "suppress_ragged_eofs": suppress_ragged_eofs,
-            "ciphers": ciphers,
-            #'ssl_version': ssl_version
+            "server_hostname": server_hostname,
         }
 
-        # TODO: We need to refactor this.
-        try:
-            sock = ssl.wrap_socket(**opts)  # pylint: disable=deprecated-method
-        except FileNotFoundError as exc:
-            # FileNotFoundError does not have missing filename info, so adding it below.
-            # Assuming that this must be ca_certs, since this is the only file path that
-            # users can pass in (`connection_verify` in the EH/SB clients) through opts above.
-            # For uamqp exception parity. Remove later when resolving issue #27128.
-            exc.filename = {"ca_certs": ca_certs}
-            raise exc
+        context = ssl.SSLContext(ssl_version)
+
+        if certfile is not None:
+            context.load_cert_chain(certfile, keyfile)
+
+        if ca_certs is not None:
+            try:
+                context.load_verify_locations(ca_certs)
+            except FileNotFoundError as exc:
+                # FileNotFoundError does not have missing filename info, so adding it below.
+                # since this is the only file path that users can pass in
+                # (`connection_verify` in the EH/SB clients) through opts above.
+                exc.filename = {"ca_certs": ca_certs}
+                raise exc from None
+
+        if ciphers is not None:
+            context.set_ciphers(ciphers)
+
         # Set SNI headers if supported
-        if (
-            (server_hostname is not None)
-            and (hasattr(ssl, "HAS_SNI") and ssl.HAS_SNI)
-            and (hasattr(ssl, "SSLContext"))
-        ):
-            context = ssl.SSLContext(opts["ssl_version"])
+        try:
+            context.check_hostname = ssl.HAS_SNI and server_hostname is not None
+        except AttributeError:
+            pass
+
+        if cert_reqs is not None:
             context.verify_mode = cert_reqs
-            if cert_reqs != ssl.CERT_NONE:
-                context.check_hostname = True
-            if (certfile is not None) and (keyfile is not None):
-                context.load_cert_chain(certfile, keyfile)
-            sock = context.wrap_socket(sock, server_hostname=server_hostname)
+
+        if ca_certs is None and context.verify_mode != ssl.CERT_NONE:
```

Comment:
can this block be moved to under line 541 in some sort of `elif` block so that the `ca_certs` logic can be grouped together?

Reply:
No I cant change the order of it as I am changing the value of `verify_mode` right above it, so I need to see what it is at the end. Setting the ssl version has some default behavior and so does the setting the value of check_hostname.

Reply:
Added a comment for clarity around the order.